### PR TITLE
Introduce proxy_env variable

### PR DESCRIPTION
**Pull request by mvdbeek**
The default proxy_env variable is empty, but users may modify their
inventory file to include global environment variables, such as
http_proxy and https_proxy.

A proxy_env variable may look like this:
```
proxy_env:
  http_proxy: http://proxy.bos.example.com:8080
  https_proxy: http://proxy.bos.example.com:8080
```

More documentation can be found in the Ansible Documentation at
http://docs.ansible.com/ansible/playbooks_environment.html

### CODE REVIEW

File: docs/faq.md [hunk 1]
```diff
@@ -20,3 +20,20 @@ galaxy_admin_pw: new_password
...
+will only see these environment variables if ansible sets these variables for
+the task. We have included a global `proxy_env` variable in the galaxy.yml playbook.
+You can set the content of this variable in your inventory. To use the proxy at
+http://proxy.bos.example.com:8080 you can define
+```
+proxy_env:
+  http_proxy: http://proxy.bos.example.com:8080
+  https_proxy: http://proxy.bos.example.com:8080
+```
+in your inventory.
```
Comment (drosofff) [inline docs/faq.md]: inventory is maybe vague here: it implicitly refers to the inventory_files dir, whereas the your modification of group_vars/all suggests that you can also define the environmental variable at group_vars level, which I think from the ansible doc is true.

Maybe we can add in commented lines in inventory files : 
```
#proxy_env:
#  http_proxy: http://proxy.bos.example.com:8080
#  https_proxy: http://proxy.bos.example.com:8080
```

**Comment by mvdbeek** [comment]
You can open a socks proxy on localhost, then point ansible to that socks proxy:
`ssh -D 8888 localhost` open a proxy on port 8888.
To confirm that is working, do
```
http_proxy=socks5://127.0.0.1:8888 https_proxy=socks5://127.0.0.1:8888 curl https://usegalaxy.org/api/version
```
which will succeed. Close the proxy (by closing the ssh session) and it will fail.
Re-open the proxy, and now use
```
proxy_env:
   http_proxy: socks5://127.0.0.1:8888
   https_proxy: socks5://127.0.0.1:8888
```
and run a task like this:

```
command: curl https://usegalaxy.org/api/version
environment: '{{ proxy_env }}'
```
When the proxy is open, it should work, when its closed it should fail.

**Comment by drosofff** [comment]
OK with the task like this...
but in the same context the GKS playbook fails because 
```
TASK [galaxyprojectdotorg.galaxy-os : Add custom Galaxy PPA (used for nginx package)] ***
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "Unsupported proxy scheme: socks5. Currently ansible only supports HTTP proxies."}
```
Thus there is no, in real, test

Maybe @colindaven can test the PR ?

File: group_vars/proxy [hunk 2]
```diff
@@ -0,0 +1,6 @@
+## Uses default settings defined in group_vars/all, but sets the `proxy_env`
+variable.
```
Comment (drosofff) [inline group_vars/proxy]: remove line break

**Comment by mvdbeek** [comment]
> Thus there is no, in real, test

You can also specify some http proxy that doesn't exist. If the task fails the environment variable has been properly propagated.

**Comment by drosofff** [comment]
@mvdbeek 

> If the task fails the environment variable has been properly propagated.

This is not exactly what is called a test... with this approach I would have more success in mines :-)

I'll see whether I can manage to get a real proxy somewhere for a test

**Comment by drosofff** [comment]
I'll give a shot to this https://gist.github.com/webinista/812c20247a6c21e639ce

**Comment by mvdbeek** [comment]
So here's my test:
```
mvandenb@u934-bdd-4-89-6580:~/src/GalaxyKickStart (doc_proxy)*$ cat hosts
[proxy]
localhost ansible_connection=local
```
```
mvandenb@u934-bdd-4-89-6580:~/src/GalaxyKickStart (doc_proxy)*$ cat proxy.yml
- hosts: all

  environment: "{{ proxy_env }}"

  tasks:
      - command: 'curl https://usegalaxy.org/api/version'
```
```
mvandenb@u934-bdd-4-89-6580:~/src/GalaxyKickStart (doc_proxy)*$ cat group_vars/proxy
## Uses default settings defined in group_vars/all, but sets the `proxy_env` variable.

proxy_env:
  http_proxy: http://www-cache:3128
  https_proxy: http://www-cache:3128
```
Now with our local proxy in place:
```
mvandenb@u934-bdd-4-89-6580:~/src/GalaxyKickStart (doc_proxy)*$ ansible-playbook -i hosts proxy.yml

PLAY [all] ***************************************************************************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************************************************************************
ok: [localhost]

TASK [command] ***********************************************************************************************************************************************************************
 [WARNING]: Consider using get_url or uri module rather than running curl

changed: [localhost]

PLAY RECAP ***************************************************************************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0
```
Remove the proxy connection and the playbook fails:
```
mvandenb@u934-bdd-4-89-6580:~/src/GalaxyKickStart (doc_proxy)*$ ansible-playbook -i hosts proxy.yml

PLAY [all] ***************************************************************************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************************************************************************
ok: [localhost]

TASK [command] ***********************************************************************************************************************************************************************
 [WARNING]: Consider using get_url or uri module rather than running curl

fatal: [localhost]: FAILED! => {"changed": true, "cmd": ["curl", "https://usegalaxy.org/api/version"], "delta": "0:00:00.024863", "end": "2017-05-12 15:30:52.329555", "failed": true, "rc": 5, "start": "2017-05-12 15:30:52.304692", "stderr": "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n                                 Dload  Upload   Total   Spent    Left  Speed\n\r  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (5) Could not resolve proxy: www-cache", "stderr_lines": ["  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current", "                                 Dload  Upload   Total   Spent    Left  Speed", "", "  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (5) Could not resolve proxy: www-cache"], "stdout": "", "stdout_lines": []}
	to retry, use: --limit @/Users/mvandenb/src/GalaxyKickStart/proxy.retry

PLAY RECAP ***************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1
```

**Comment by drosofff** [comment]
I have a squid proxy in the cloud and an ansible target vm running behind the proxy vm.
@mvdbeek although your test works perfectly under these settings there is still errors to update the apt cache in with galaxy-os role.
I may merge this PR, but we need a real test to the playbook under a proxy

**Comment by drosofff** [comment]
So here is the progress with a squid proxy open on port 3128
group_vars/proxy:
```
proxy_env:
  http_proxy: http://instance-1:3128
  https_proxy: http://instance-1:3128
```
and inventory_files/galaxy-kickstart
```
[kickstart]
localhost ansible_connection=local
[proxy]
localhost ansible_connection=local
```
Then,
Adding
```
# Install system packages
- name: Add Ubuntu repository key
  apt_key: keyserver=hkp://keyserver.ubuntu.com:80 id=E9708E5E0DEE60262A3FE2089EB12D87B9BDC5ED
```
in roles/galaxyprojectdotorg.galaxy-os/tasks/ubuntu/packages.yml

allows now to go up to the galaxy-tools role, which fails at: 

```
TASK [galaxyprojectdotorg.galaxy-tools : Create Galaxy bootstrap user] *********
fatal: [localhost]: FAILED! => {"changed": true, "cmd": ["/home/galaxy/galaxy/.venv/bin/python", "manage_bootstrap_user.py", "-c", "/home/galaxy/galaxy/config/galaxy.ini", "create", "-e", "admin@galaxy.org", "-u", "cloud", "-p", "admin", "-a", "admin"], "delta": "0:00:00.209513", "end": "2017-05-13 20:39:47.989883", "failed": true, "rc": 1, "start": "2017-05-13 20:39:47.780370", "stderr": "Traceback (most recent call last):\n  File \"manage_bootstrap_user.py\", line 19, in <module>\n    import galaxy.config\n  File \"/home/galaxy/galaxy/lib/galaxy/config.py\", line 26, in <module>\n    from galaxy.util.postfork import register_postfork_function\n  File \"/home/galaxy/galaxy/lib/galaxy/util/postfork.py\", line 10, in <module>\n    if uwsgi.numproc:\nAttributeError: 'module' object has no attribute 'numproc'", "stdout": "", "stdout_lines": [], "warnings": []}
```
Its ok without proxy but after trying with proxy, reverting to no-proxy does not revert this error...

**Comment by drosofff** [comment]
OK, I figured out the problem.

@mvdbeek one have to had in group_vars/proxy:

```
## Uses default settings defined in group_vars/all, but sets the `proxy_env` variable.

proxy_env:
  http_proxy: http://proxy.bos.example.com:8080
  https_proxy: http://proxy.bos.example.com:8080
  no_proxy: localhost,127.0.0.0,127.0.1.1,127.0.1.1,local.home
```

**Comment by mvdbeek** [comment]
Ah, that makes sense. I'll add that

**Comment by drosofff** [comment]
👍 
For records, the squid docker container used to test with proxy was https://github.com/sameersbn/docker-squid

**Pull request merged**